### PR TITLE
cdc/sink: fix kafka consumer update globalResolvedTs too early, and adjust ddl sink processing logic.

### LINKED ISSUE #4845: cdc should send ddl event before send checkpoint.
### What did you do?

unstable kafka integration test reports the problem.

When the `checkpointTs == DDL.CommitTs` is true, the DDL should be sent first, but the checkpoinTs was sent first.

### What did you expect to see?

send DDL first, then checkpointTs.

### What did you see instead?

```
[2022/03/10 17:24:12.846 +08:00] [DEBUG] [mq.go:220] ["emit checkpointTs"] [checkpointTs=431726066541264898]
[2022/03/10 17:24:13.491 +08:00] [DEBUG] [mq.go:258] ["emit ddl event"] [query="CREATE TABLE `attributes_t2` (`id` INT PRIMARY KEY,`name` VARCHAR(50)) PARTITION BY RANGE (`id`) (PARTITION `p0` VALUES LESS THAN (10000),PARTITION `p1` VALUES LESS THAN (MAXVALUE))"] [commitTs=431726066541264898] [partition=-1] [changefeed=cf172ec0-1b46-4592-af9e-447a9f1f4c0d] [role=owner]
```

### Versions of the cluster

Upstream TiDB cluster version (execute `SELECT tidb_version();` in a MySQL client):

```console
master
```

Upstream TiKV version (execute `tikv-server --version`):

```console
master
```

TiCDC version (execute `cdc version`):

```console
master
```

### PR DESCRIPTION
**Pull request by 3AceShowHand**
### What problem does this PR solve?

Issue Number: close #4845, #4731

### What is changed and how it works?

* Kafka consumer does not need to check DDL's CommitTs when receiving a new one, before emitting the checkpoinTs, emit DDL first.

### Check List 

Tests 

 - No code

Code changes

Side effects

Related changes

 - Need to cherry-pick to the release branch

### Release note 

```release-note
None
```

### CODE REVIEW

**Comment by 3AceShowHand** [comment]
```

[2022/03/08 20:55:22.748 +08:00] [DEBUG] [mq.go:220] ["emit checkpointTs"] [checkpointTs=431684088547508243]
[2022/03/08 20:55:22.750 +08:00] [DEBUG] [etcd_worker.go:454] ["[etcd worker] ==========Update State to ETCD=========="]
[2022/03/08 20:55:22.750 +08:00] [DEBUG] [etcd_worker.go:459] ["[etcd worker] put key"] [key=/tidb/cdc/task/status/9127383f-42a9-48cd-a1b6-2f6a05772fdf/92a8d592-2fb6-4f01-b9b7-452887cd04c6] [value="{\"tables\":{\"67\":{\"start-ts\":431684087407181842,\"mark-table-id\":0},\"69\":{\"start-ts\":431684088508186628,\"mark-table-id\":0},\"71\":{\"start-ts\":431684088521555992,\"mark-table-id\":0},\"73\":{\"start-ts\":431684088547508243,\"mark-table-id\":0}},\"operation\":{\"73\":{\"delete\":false,\"boundary_ts\":431684088547508243,\"status\":2}},\"admin-job-type\":0}"]
[2022/03/08 20:55:22.750 +08:00] [DEBUG] [etcd_worker.go:462] ["[etcd worker] ============State Commit============="] [committed=true]
[2022/03/08 20:55:22.752 +08:00] [DEBUG] [mq.go:258] ["emit ddl event"] [query="CREATE TABLE `tp_text` (`id` INT AUTO_INCREMENT,`c_tinytext` TINYTEXT NULL,`c_text` TEXT NULL,`c_mediumtext` MEDIUMTEXT NULL,`c_longtext` LONGTEXT NULL,PRIMARY KEY `pk`(`id`))"] [commitTs=431684088547508243] [partition=-1] [changefeed=92a8d592-2fb6-4f01-b9b7-452887cd04c6] [role=owner]
```

```
[2022/03/08 20:55:22.754 +08:00] [INFO] [main.go:590] ["update globalResolvedTs"] [ts=431684088547508243]
[2022/03/08 20:55:22.754 +08:00] [WARN] [main.go:496] ["ignore redundant ddl event, commitTs == globalResolvedTs"] [ddl="{\"StartTs\":0,\"CommitTs\":431684088547508243,\"TableInfo\":{\"Schema\":\"row_format\",\"Table\":\"tp_text\",\"TableID\":0,\"ColumnInfo\":null},\"PreTableInfo\":null,\"Query\":\"CREATE TABLE `tp_text` (`id` INT AUTO_INCREMENT,`c_tinytext` TINYTEXT NULL,`c_text` TEXT NULL,`c_mediumtext` MEDIUMTEXT NULL,`c_longtext` LONGTEXT NULL,PRIMARY KEY `pk`(`id`))\",\"Type\":3}"]
[2022/03/08 20:55:22.756 +08:00] [WARN] [main.go:496] ["ignore redundant ddl event, commitTs == globalResolvedTs"] [ddl="{\"StartTs\":0,\"CommitTs\":431684088547508243,\"TableInfo\":{\"Schema\":\"row_format\",\"Table\":\"tp_text\",\"TableID\":0,\"ColumnInfo\":null},\"PreTableInfo\":null,\"Query\":\"CREATE TABLE `tp_text` (`id` INT AUTO_INCREMENT,`c_tinytext` TINYTEXT NULL,`c_text` TEXT NULL,`c_mediumtext` MEDIUMTEXT NULL,`c_longtext` LONGTEXT NULL,PRIMARY KEY `pk`(`id`))\",\"Type\":3}"]
[2022/03/08 20:55:22.757 +08:00] [WARN] [main.go:496] ["ignore redundant ddl event, commitTs == globalResolvedTs"] [ddl="{\"StartTs\":0,\"CommitTs\":431684088547508243,\"TableInfo\":{\"Schema\":\"row_format\",\"Table\":\"tp_text\",\"TableID\":0,\"ColumnInfo\":null},\"PreTableInfo\":null,\"Query\":\"CREATE TABLE `tp_text` (`id` INT AUTO_INCREMENT,`c_tinytext` TINYTEXT NULL,`c_text` TEXT NULL,`c_mediumtext` MEDIUMTEXT NULL,`c_longtext` LONGTEXT NULL,PRIMARY KEY `pk`(`id`))\",\"Type\":3}"]
[2022/03/08 20:55:22.757 +08:00] [WARN] [main.go:496] ["ignore redundant ddl event, commitTs == globalResolvedTs"] [ddl="{\"StartTs\":0,\"CommitTs\":431684088547508243,\"TableInfo\":{\"Schema\":\"row_format\",\"Table\":\"tp_text\",\"TableID\":0,\"ColumnInfo\":null},\"PreTableInfo\":null,\"Query\":\"CREATE TABLE `tp_text` (`id` INT AUTO_INCREMENT,`c_tinytext` TINYTEXT NULL,`c_text` TEXT NULL,`c_mediumtext` MEDIUMTEXT NULL,`c_longtext` LONGTEXT NULL,PRIMARY KEY `pk`(`id`))\",\"Type\":3}"]
[2022/03/08 20:55:26.754 +08:00] [INFO] [main.go:590] ["update globalResolvedTs"] [ts=431684088560615447]
[2022/03/08 20:55:26.756 +08:00] [ERROR] [mysql.go:722] ["execute DMLs failed"] [err="[CDC:ErrMySQLTxnError]MySQL txn error: Error 1146: Table 'row_format.tp_text' doesn't exist"]
[2022/03/08 20:55:26.756 +08:00] [PANIC] [main.go:272] ["Error running consumer"] [error="[CDC:ErrMySQLTxnError]MySQL txn error: Error 1146: Table 'row_format.tp_text' doesn't exist"] [errorVerbose="[CDC:ErrMySQLTxnError]MySQL txn error: Error 1146: Table 'row_format.tp_text' doesn't exist\ngithub.com/pingcap/errors.AddStack\n\tgithub.com/pingcap/errors@v0.11.5-0.20211224045212-9687c2b0f87c/errors.go:174\ngithub.com/pingcap/errors.(*Error).GenWithStackByArgs\n\tgithub.com/pingcap/errors@v0.11.5-0.20211224045212-9687c2b0f87c/normalize.go:164\ngithub.com/pingcap/tiflow/pkg/errors.WrapError\n\tgithub.com/pingcap/tiflow/pkg/errors/helper.go:30\ngithub.com/pingcap/tiflow/cdc/sink.(*mysqlSink).execDMLWithMaxRetries.func1.3\n\tgithub.com/pingcap/tiflow/cdc/sink/mysql.go:571\ngithub.com/pingcap/tiflow/cdc/sink.(*Statistics).RecordBatchExecution\n\tgithub.com/pingcap/tiflow/cdc/sink/statistics.go:109\ngithub.com/pingcap/tiflow/cdc/sink.(*mysqlSink).execDMLWithMaxRetries.func1\n\tgithub.com/pingcap/tiflow/cdc/sink/mysql.go:558\ngithub.com/pingcap/tiflow/pkg/retry.run\n\tgithub.com/pingcap/tiflow/pkg/retry/retry_with_opt.go:54\ngithub.com/pingcap/tiflow/pkg/retry.Do\n\tgithub.com/pingcap/tiflow/pkg/retry/retry_with_opt.go:32\ngithub.com/pingcap/tiflow/cdc/sink.(*mysqlSink).execDMLWithMaxRetries\n\tgithub.com/pingcap/tiflow/cdc/sink/mysql.go:550\ngithub.com/pingcap/tiflow/cdc/sink.(*mysqlSink).execDMLs\n\tgithub.com/pingcap/tiflow/cdc/sink/mysql.go:721\ngithub.com/pingcap/tiflow/cdc/sink.(*mysqlSinkWorker).run.func3\n\tgithub.com/pingcap/tiflow/cdc/sink/mysql_worker.go:114\ngithub.com/pingcap/tiflow/cdc/sink.(*mysqlSinkWorker).run\n\tgithub.com/pingcap/tiflow/cdc/sink/mysql_worker.go:150\ngithub.com/pingcap/tiflow/cdc/sink.(*mysqlSink).createSinkWorkers.func1\n\tgithub.com/pingcap/tiflow/cdc/sink/mysql.go:392\nruntime.goexit\n\truntime/asm_amd64.s:1371"] [stack="main.main.func2\n\tcommand-line-arguments/main.go:272"]
```

CDC emit checkpointTs before DDL, but the consumer receive DDL before checkpointTs, we need to investigate this problem.

**Comment by 3AceShowHand** [comment]
There might have multiple DDL events with the same CommitTs.

**Comment by 3AceShowHand** [comment]
```
// changefeed-test-2 ddlFinishedTs=431833944692621330
[2022/03/15 11:42:51.858 +08:00] [INFO] [ddl_sink.go:162] ["emit ddl event"] [DDL="{\"StartTs\":431833944692621328,\"CommitTs\":431833944692621330,\"TableInfo\":{\"Schema\":\"test\",\"Table\":\"simple2\",\"TableID\":67,\"ColumnInfo\":[{\"Name\":\"id\",\"Type\":3},{\"Name\":\"val\",\"Type\":3}]},\"PreTableInfo\":null,\"Query\":\"CREATE TABLE `test`.`simple2` (`id` INT PRIMARY KEY,`val` INT)\",\"Type\":3}"]
[2022/03/15 11:42:51.858 +08:00] [DEBUG] [black_hole.go:75] ["BlockHoleSink: DDL Event"] [ddl="{\"StartTs\":431833944692621328,\"CommitTs\":431833944692621330,\"TableInfo\":{\"Schema\":\"test\",\"Table\":\"simple2\",\"TableID\":67,\"ColumnInfo\":[{\"Name\":\"id\",\"Type\":3},{\"Name\":\"val\",\"Type\":3}]},\"PreTableInfo\":null,\"Query\":\"CREATE TABLE `test`.`simple2` (`id` INT PRIMARY KEY,`val` INT)\",\"Type\":3}"]
[2022/03/15 11:42:51.858 +08:00] [INFO] [ddl_sink.go:168] ["Execute DDL succeeded"] [changefeed=changefeed-test2] [ignored=false] [ddl="{\"StartTs\":431833944692621328,\"CommitTs\":431833944692621330,\"TableInfo\":{\"Schema\":\"test\",\"Table\":\"simple2\",\"TableID\":67,\"ColumnInfo\":[{\"Name\":\"id\",\"Type\":3},{\"Name\":\"val\",\"Type\":3}]},\"PreTableInfo\":null,\"Query\":\"CREATE TABLE `test`.`simple2` (`id` INT PRIMARY KEY,`val` INT)\",\"Type\":3}"]
[2022/03/15 11:42:51.945 +08:00] [INFO] [ddl_sink.go:219] ["ddl already executed"] [ddlFinishedTs=431833944692621330] [DDL="{\"StartTs\":431833944692621328,\"CommitTs\":431833944692621330,\"TableInfo\":{\"Schema\":\"test\",\"Table\":\"simple2\",\"TableID\":67,\"ColumnInfo\":[{\"Name\":\"id\",\"Type\":3},{\"Name\":\"val\",\"Type\":3}]},\"PreTableInfo\":null,\"Query\":\"CREATE TABLE `test`.`simple2` (`id` INT PRIMARY KEY,`val` INT)\",\"Type\":3}"] [changefeed=changefeed-test2]
...
[2022/03/15 11:42:57.163 +08:00] [INFO] [changefeed.go:261] ["initialize changefeed"] [changefeed=changefeed-test2] [info="{\"sink-uri\":\"blackhole:\",\"opts\":{},\"create-time\":\"2022-03-15T11:42:43.150217418+08:00\",\"start-ts\":431833943355686914,\"target-ts\":0,\"admin-job-type\":0,\"sort-engine\":\"unified\",\"sort-dir\":\"\",\"config\":{\"case-sensitive\":true,\"enable-old-value\":true,\"force-replicate\":false,\"check-gc-safe-point\":true,\"filter\":{\"rules\":[\"*.*\"],\"ignore-txn-start-ts\":null},\"mounter\":{\"worker-num\":32},\"sink\":{\"dispatchers\":null,\"protocol\":\"\",\"column-selectors\":null},\"cyclic-replication\":{\"enable\":false,\"replica-id\":0,\"filter-replica-ids\":null,\"id-buckets\":0,\"sync-ddl\":false},\"scheduler\":{\"type\":\"table-number\",\"polling-time\":-1},\"consistent\":{\"level\":\"none\",\"max-log-size\":64,\"flush-interval\":1000,\"storage\":\"\"}},\"state\":\"normal\",\"error\":null,\"sync-point-enabled\":false,\"sync-point-interval\":600000000000,\"creator-version\":\"v5.4.0-master\"}"] ["checkpoint ts"=431833944692621330]
[2022/03/15 11:42:57.460 +08:00] [DEBUG] [processor.go:1050] ["finished gc in schema storage"] [gcTs=431833944692621330] [changefeed=changefeed-test2]
[2022/03/15 11:42:57.462 +08:00] [INFO] [puller.go:215] ["puller is initialized"] [duration=6.743079ms] [changefeed=changefeed-test2] [tableID=-1] [spans="[\"[6d44444c4a6f624cff69ff737400000000ff0000f90000000000ff00006c0000000000fa, 6d44444c4a6f624cff69ff737400000000ff0000f90000000000ff00006d0000000000fa)\",\"[6d44444c4a6f6241ff64ff644964784c69ff7374ff0000000000ff000000f700000000ff0000006c00000000fb, 6d44444c4a6f6241ff64ff644964784c69ff7374ff0000000000ff000000f700000000ff0000006d00000000fb)\"]"] [resolvedTs=431833944692621330]
[2022/03/15 11:42:59.249 +08:00] [INFO] [puller.go:215] ["puller is initialized"] [duration=5.476986ms] [changefeed=changefeed-test2] [tableID=65] [spans="[\"[7480000000000000ff415f720000000000fa, 7480000000000000ff415f730000000000fa)\"]"] [resolvedTs=431833944692621330]
...
[2022/03/15 11:43:01.145 +08:00] [INFO] [ddl_sink.go:225] ["ddl is not finished yet"] [ddlSentTs=431833944692621330] [ddlFinishedTs=0] [DDL="{\"StartTs\":431833944692621328,\"CommitTs\":431833944692621330,\"TableInfo\":{\"Schema\":\"test\",\"Table\":\"simple2\",\"TableID\":67,\"ColumnInfo\":[{\"Name\":\"id\",\"Type\":3},{\"Name\":\"val\",\"Type\":3}]},\"PreTableInfo\":null,\"Query\":\"CREATE TABLE `test`.`simple2` (`id` INT PRIMARY KEY,`val` INT)\",\"Type\":3}"] [changefeed=changefeed-test2]
[2022/03/15 11:43:01.278 +08:00] [PANIC] [ddl_sink.go:157] ["DDL regression, please report a bug"] [checkpointTs=431833944692621330] [DDL="{\"StartTs\":431833944692621328,\"CommitTs\":431833944692621330,\"TableInfo\":{\"Schema\":\"test\",\"Table\":\"simple2\",\"TableID\":67,\"ColumnInfo\":[{\"Name\":\"id\",\"Type\":3},{\"Name\":\"val\",\"Type\":3}]},\"PreTableInfo\":null,\"Query\":\"CREATE TABLE `test`.`simple2` (`id` INT PRIMARY KEY,`val` INT)\",\"Type\":3}"] [stack="github.com/pingcap/tiflow/cdc/owner.(*ddlSinkImpl).run.func1\n\t/home/jenkins/agent/workspace/cdc_ghpr_kafka_integration_test/go/src/github.com/pingcap/tiflow/cdc/owner/ddl_sink.go:157"]
```

it looks that redundant DDL which has commitTs equal to the current checkpointTs would be sent multiple times.

File: cdc/owner/changefeed.go [hunk 1]
```diff
@@ -207,10 +207,9 @@ func (c *changefeed) tick(ctx cdcContext.Context, state *orchestrator.Changefeed
 		// This condition implies that the DDL resolved-ts has not yet reached checkpointTs,
 		// which implies that it would be premature to schedule tables or to update status.
 		// So we return here.
-		log.Info("barrierTs < checkpointTs, premature to schedule tables or update status",
```
Comment (sdojjy) [inline cdc/owner/changefeed.go:210]: why delete this info log?

File: cmd/kafka-consumer/main.go [hunk 2]
```diff
@@ -502,8 +502,9 @@ func (c *Consumer) ConsumeClaim(session sarama.ConsumerGroupSession, claim saram
 						zap.Uint64("ts", ts),
 						zap.Uint64("resolvedTs", resolvedTs),
 						zap.Int32("partition", partition))
-				} else if ts > resolvedTs {
-					log.Debug("update sink resolved ts",
+				}
+				if ts > resolvedTs {
```
Comment (sdojjy) [inline cmd/kafka-consumer/main.go:506]: so, here no logic is changed, only change the log level?

File: cdc/owner/changefeed.go (hunk 1 above)
Comment (3AceShowHand) [reply cdc/owner/changefeed.go:210]: it's noisy, and may useless.

File: cmd/kafka-consumer/main.go [hunk 3]
```diff
@@ -574,9 +579,20 @@ func (c *Consumer) forEachSink(fn func(sink *partitionSink) error) error {
 	return nil
 }
 
-// Run runs the Consumer
+func (c *Consumer) getMinPartitionResolvedTs() (result uint64, err error) {
+	result = uint64(math.MaxUint64)
+	err = c.forEachSink(func(sink *partitionSink) error {
+		a := atomic.LoadUint64(&sink.resolvedTs)
```
Comment (sdojjy) [inline cmd/kafka-consumer/main.go:585]: change `a`  to a readable name？

File: cmd/kafka-consumer/main.go (hunk 2 above)
Comment (3AceShowHand) [reply cmd/kafka-consumer/main.go:506]: yes, for better observe the resolved ts update logic.

File: cmd/kafka-consumer/main.go (hunk 3 above)
Comment (3AceShowHand) [reply cmd/kafka-consumer/main.go:585]: it's only a local variable, does not affect out scope, right ?

**Review by sdojjy — APPROVED** [review]

File: cdc/owner/ddl_sink.go [hunk 4]
```diff
@@ -148,6 +148,13 @@ func (s *ddlSinkImpl) run(ctx cdcContext.Context, id model.ChangeFeedID, info *m
 			case err := <-s.errCh:
 				ctx.Throw(err)
 				return
+			default:
+			}
+
+			// `ticker.C` and `ddlCh` may can be triggered at the same time, it
+			// does not matter which one emit first, since TiCDC allow DDL with
+			// CommitTs equal to the last CheckpointTs be emitted later.
+			select {
```
Comment (overvenus) [inline cdc/owner/ddl_sink.go:157]: I think we should also select errCh here.

Comment (overvenus) [reply cdc/owner/ddl_sink.go:156]: Should also select `ctx.Done()`.

**Review by overvenus — APPROVED** [review]
Rest LGTM

**Pull request merged**